### PR TITLE
GNNModular

### PR DESCRIPTION
**Pull request by tonydavis629**
This PR adds GNNModular to torch_models. GNNModular allows users to define graph convolutional encoders, prediction heads, and specific tasks such as node prediction, edge prediction, regression, classification, etc. This PR implements edge prediction only, using the GINEConv layer from torch geometric. as the GCN

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/gnn.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    >>> features = featurizer.featurize(smiles)
+    >>> batched_graph = BatchGraphData(features).numpy_to_torch(device="cuda")
+    >>> modular = model = GNNModular("gin", 3, 64, 1, "attention", 0, "last", "edge_pred")
+    >>> gnnmodel = modular.gnn
+    >>> print(gnnmodel(batched_graph)[0].shape)
+    torch.Size([23, 64])
+
+    """
+
+    def __init__(self,
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:37]: Can you document the parameters in the docstring?

File: deepchem/models/torch_models/gnn.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                 atom_type_embedding,
+                 chirality_embedding,
+                 gconvs,
+                 batch_norms,
+                 dropout,
+                 JK,
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:43]: What does JK mean? Generally better to use descriptive rather than short variable names

File: deepchem/models/torch_models/gnn.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+            h_list = [h.unsqueeze_(0) for h in h_list]
+            node_representation = torch.max(torch.cat(h_list, dim=0), dim=0)[0]
+        elif self.JK == "sum":
+            h_list = [h.unsqueeze_(0) for h in h_list]
+            node_representation = torch.sum(torch.cat(h_list, dim=0), dim=0)[0]
+
+        return node_representation, data
+
+
+class GNN_head(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:96]: This should be camel case since it's a class name. So `GNNHead`

File: deepchem/models/torch_models/gnn.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    data: BatchGraphData
+        The input graph data.
+
+    Returns
+    -------
+    out: torch.Tensor
+        The output of the GNN head module.
+    """
+
+    def __init__(self, pool, head):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:113]: Can you document the parameters for the init? The parameters documented above are only for the forward pass

File: deepchem/models/torch_models/gnn.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        self.dropout = dropout
+        self.num_layer = len(gconvs)
+        self.JK = JK
+
+        # may mess with loading pretrained weights
+        if init_emb:
+            torch.nn.init.xavier_uniform_(self.atom_type_embedding.weight.data)
+            torch.nn.init.xavier_uniform_(self.chirality_embedding.weight.data)
+
+    def forward(self, data: BatchGraphData):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:60]: Can you add a docstring for the `forward` argments?

File: deepchem/models/torch_models/gnn.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        for layer in range(self.num_layer):
+            if self.gnn_type == "gin":
+                encoders.append(
+                    GINEConv(
+                        torch.nn.Linear(self.emb_dim, self.emb_dim),
+                        edge_dim=2,  # bond type, bond direction
+                        aggr="add"))
+            else:
+                raise ValueError("Only GIN is supported for now")
+            # elif self.gnn_type == "gcn":
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:226]: Looks like some leftover cruft

File: deepchem/models/torch_models/gnn.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        if self.task == "edge_pred":  # unsupervised task, does not need pred head
+            return self.gnn
+        else:
+            return torch.nn.Sequential(self.gnn, self.gnn_head)
+
+    def loss_func(self, inputs, labels, weights):
+        if self.task == "edge_pred":
+            return self.edge_pred_loss(inputs, labels, weights)
+
+    def edge_pred_loss(self, inputs, labels, weights):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:296]: Can you add a docstring here?

File: deepchem/models/torch_models/gnn.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+        self.gnn_head = GNN_head(components['pool'], components['head'])
+        return components
+
+    def build_model(self):
+        if self.task == "edge_pred":  # unsupervised task, does not need pred head
+            return self.gnn
+        else:
+            return torch.nn.Sequential(self.gnn, self.gnn_head)
+
+    def loss_func(self, inputs, labels, weights):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:292]: Can you add a docstring here?

File: deepchem/models/torch_models/gnn.py [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+            'head':
+                head
+        }
+        self.gnn = GNN(components['atom_type_embedding'],
+                       components['chirality_embedding'], components['gconvs'],
+                       components['batch_norms'], self.dropout, self.JK)
+        self.gnn_head = GNN_head(components['pool'], components['head'])
+        return components
+
+    def build_model(self):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:286]: Can you add a docstring here?

File: deepchem/models/torch_models/gnn.py [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+                                   dim=1)
+
+        loss = self.criterion(
+            positive_score, torch.ones_like(positive_score)) + self.criterion(
+                negative_score, torch.zeros_like(negative_score))
+        return (loss * weights[0]).mean()
+
+    def _prepare_batch(self, batch):
+        """
+        Prepares the batch for the model by converting the GraphData numpy arrays to torch tensors and moving them to the device.
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:314]: Can you add a Parameters section here?

File: deepchem/models/torch_models/gnn.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+        _, labels, weights = super()._prepare_batch(([], labels, weights))
+
+        if (len(labels) != 0) and (len(weights) != 0):
+            labels = labels[0]
+            weights = weights[0]
+
+        return inputs, labels, weights
+
+
+class NegativeEdge:
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:329]: Following up on our offline conversation, this should probably be in utils as we discussed

File: deepchem/models/torch_models/gnn.py [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+    """
+    NegativeEdge is a callable class that adds negative edges to the input graph data. It randomly samples negative edges (edges that do not exist in the original graph) and adds them to the input graph data.
+    The number of negative edges added is equal to half the number of edges in the original graph. This is useful for tasks like edge prediction, where the model needs to learn to differentiate between existing and non-existing edges.
+    """
+
+    def __call__(self, data):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:335]: Can you add a docstring?

File: deepchem/models/torch_models/gnn.py [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+            labels = labels[0]
+            weights = weights[0]
+
+        return inputs, labels, weights
+
+
+class NegativeEdge:
+    """
+    NegativeEdge is a callable class that adds negative edges to the input graph data. It randomly samples negative edges (edges that do not exist in the original graph) and adds them to the input graph data.
+    The number of negative edges added is equal to half the number of edges in the original graph. This is useful for tasks like edge prediction, where the model needs to learn to differentiate between existing and non-existing edges.
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:332]: Can you add a usage example?

File: deepchem/models/torch_models/modular.py [hunk 14]
```diff
@@ -84,7 +84,7 @@ def __init__(self, model: nn.Module, components: dict, **kwargs):
         super().__init__(self.model, self.loss_func, **kwargs)
         self.model.to(self.device)
         self.components = {
-            k: v.to(self.device) for k, v in self.components.items()
+            k: v.to(self.device) for k, v in self.components.items() if isinstance(v, nn.Module)
```
Comment (rbharath) [inline deepchem/models/torch_models/modular.py:87]: When would we expect to have a non `nn.module` in self.components?

File: deepchem/models/torch_models/gnn.py [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+from deepchem.feat.graph_data import BatchGraphData
+
+num_atom_type = 120
+num_chirality_tag = 3
+# Relevant in future PRs
+# num_bond_type = 6
+# num_bond_direction = 3
+
+
+class GNN(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:15]: Can you add this to `docs/source/api_reference/layers.rst`?

File: deepchem/models/torch_models/gnn.py [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+        elif self.JK == "max":
+            h_list = [h.unsqueeze_(0) for h in h_list]
+            node_representation = torch.max(torch.cat(h_list, dim=0), dim=0)[0]
+        elif self.JK == "sum":
+            h_list = [h.unsqueeze_(0) for h in h_list]
+            node_representation = torch.sum(torch.cat(h_list, dim=0), dim=0)[0]
+
+        return node_representation, data
+
+
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:95]: Can you add this class to `docs/source/api_reference/layers.rst`?

**Review by rbharath — COMMENTED** [review]
Good start! Mostly requests for more documentation below

File: deepchem/models/torch_models/gnn.py (hunk 6 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:226]: They will be relevant in the next PR

File: deepchem/models/torch_models/modular.py (hunk 14 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/modular.py:87]: In this GNNModular class. The pooling components can be either a function or parameterized nn.module.

File: deepchem/models/tests/test_gnn.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+                               featurizer=featurizer)
+    dataset = loader.create_dataset(input_file)
+    metric = dc.metrics.Metric(dc.metrics.mean_absolute_error,
+                               mode="regression")
+
+    return dataset, metric
+
+
+@pytest.mark.torch
+def test_GNN_edge_pred():
```
Comment (rbharath) [inline deepchem/models/tests/test_gnn.py:25]: We will need the usual overfit, save/reload tests. Will those follow in subsequent PRs?

File: deepchem/models/torch_models/gnn.py (hunk 2 above)
Comment (rbharath) [reply deepchem/models/torch_models/gnn.py:68]: Can you replace with `jump_knowledge`? `JK` is shorter but less maintainable and readable

File: deepchem/models/torch_models/gnn.py (hunk 4 above)
Comment (rbharath) [reply deepchem/models/torch_models/gnn.py:143]: Quick reminder here. The arguments documented above are for `forward` and not `__init__`

**Review by rbharath — COMMENTED** [review]
One minor refactoring request below (`self.JK` -> `self.jump_knowledge`). Once that's done and a requested comment below is added this should be ready to merge in

File: deepchem/models/torch_models/gnn.py (hunk 4 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:143]: Sorry got confused with the other PRs. This is fixed.

File: deepchem/models/torch_models/gnn.py (hunk 3 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py]: fixed

File: deepchem/models/torch_models/gnn.py (hunk 2 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:68]: fixed

File: deepchem/models/torch_models/gnn.py (hunk 10 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py]: fixed

File: deepchem/models/torch_models/gnn.py (hunk 11 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py]: modified as a helper function

File: deepchem/models/torch_models/gnn.py (hunk 12 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py]: done

File: deepchem/models/tests/test_gnn.py (hunk 17 above)
Comment (tonydavis629) [reply deepchem/models/tests/test_gnn.py:25]: test_GNN_edge_pred() is my attempt at a low resource overfit test, just checking if the loss is going down after between 5 epochs. Hard to determine what 'overfit' means for an unsupervised task. I have added a save/reload test. There are more tests in the coming PRs.

**Review by rbharath — APPROVED** [review]
LGTM, feel free to merge once CI is clear
